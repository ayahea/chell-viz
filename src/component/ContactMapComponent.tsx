import * as React from 'react';
import { CartesianGrid, Dot, ReferenceLine, Scatter, ScatterChart, XAxis, YAxis, ZAxis } from 'recharts';

import { CONTACT_MAP_DATA_TYPE, ICouplingScore } from 'chell';
import { withDefaultProps } from '../helper/ReactHelper';
import { ChellSlider } from './ChellSlider';

export type CONTACT_MAP_CB_RESULT_TYPE = ICouplingScore;
export type ContactMapCallback = (coupling: CONTACT_MAP_CB_RESULT_TYPE) => void;

const defaultProps = {
  data: {
    contactMonomer: [],
    couplingScore: [],
    distanceMapMonomer: [],
  } as CONTACT_MAP_DATA_TYPE,
  onClick: undefined as ContactMapCallback | undefined,
  onMouseEnter: undefined as ContactMapCallback | undefined,
  selectedData: undefined as number | undefined,
};
const initialState = { min_x: 1000, max_x: 0, nodeSize: 2, probabilityFilter: 0.99 };

type Props = {} & typeof defaultProps;
type State = Readonly<typeof initialState>;

export const ContactMapComponent = withDefaultProps(
  defaultProps,
  class ContactMapComponentClass extends React.Component<Props, State> {
    public readonly state: State = initialState;

    constructor(props: Props) {
      super(props);
    }

    public componentDidUpdate(prevProps: Props, prevState: State) {
      const { data } = this.props;
      if (data !== prevProps.data) {
        let max = initialState.max_x;
        let min = initialState.min_x;
        for (const contact of data.contactMonomer) {
          max = Math.max(max, contact.i);
          min = Math.min(min, contact.i);
        }
        this.setState({
          max_x: max,
          min_x: min,
        });
      }
    }

    public render() {
      const { data, selectedData } = this.props;
      const domain = [Math.max(0, this.state.min_x - 5), this.state.max_x + 5];
      const blackDots = new Array<ICouplingScore>();
      data.couplingScore.filter(coupling => coupling.probability >= this.state.probabilityFilter).forEach(coupling => {
        blackDots.push(coupling);
        blackDots.push({
          ...coupling,
          i: coupling.j,
          // tslint:disable-next-line:object-literal-sort-keys
          A_i: coupling.A_j,
          j: coupling.i,
          A_j: coupling.A_i,
        });
      });
      return data ? (
        <div>
          <ScatterChart width={400} height={400}>
            <XAxis type="number" dataKey={'i'} orientation={'top'} domain={domain} />
            <YAxis type="number" dataKey={'j'} reversed={true} domain={domain} />
            <ZAxis type="number" dataKey={'probability'} />
            {selectedData && <ReferenceLine x={selectedData} stroke={'#ff0000'} />}
            {selectedData && <ReferenceLine y={selectedData} stroke={'#ff0000'} />}
            <CartesianGrid />
            <Scatter
              name="contacts_monomer"
              data={data.contactMonomer}
              fill="#009999"
              onClick={this.onClick()}
              shape={<Dot r={this.state.nodeSize} />}
            />
            <Scatter
              name="CouplingScoresCompared"
              data={blackDots}
              fill="#000000"
              onClick={this.onClick()}
              onMouseEnter={this.onMouseEnter()}
              shape={<Dot r={this.state.nodeSize} />}
            />
          </ScatterChart>
          <ChellSlider
            max={100}
            min={0}
            label={'Probability'}
            defaultValue={99}
            onChange={this.onProbabilityChange()}
          />
          <ChellSlider
            max={10}
            min={1}
            label={'Node Size'}
            defaultValue={this.state.nodeSize * 2}
            onChange={this.onNodeSizeChange()}
          />
        </div>
      ) : null;
    }

    protected onClick = () => (coupling: ICouplingScore) => {
      if (this.props.onClick) {
        this.props.onClick(coupling);
      }
    };

    protected onProbabilityChange = () => (value: number) => {
      this.setState({
        probabilityFilter: value / 100,
      });
    };

    protected onNodeSizeChange = () => (value: number) => {
      this.setState({
        nodeSize: value / 2,
      });
    };

    protected onMouseEnter = () => (e: { [key: string]: any; payload: ICouplingScore }) => {
      const { payload } = e;
      const { onMouseEnter } = this.props;

      if (onMouseEnter) {
        onMouseEnter(payload);
      }
    };

    /*
  Exposed recharts mouse events to potentially become props:
  - onMouseDown
  - onMouseUp
  - onMouseMove
  - onMouseOut
  - onMouseOver
  - onMouseLeave
  */
  },
);
