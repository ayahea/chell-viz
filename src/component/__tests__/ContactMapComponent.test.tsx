import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as React from 'react';

import { ContactMapComponent } from '../ContactMapComponent';

describe('ContactMapComponent', () => {
  const emptyData = {
    contactMonomer: [],
    couplingScore: [],
    distanceMapMonomer: [],
  };

  const sampleData = {
    contactMonomer: [{ i: 0, j: 1, dist: 10 }, { i: 1, j: 0, dist: 10 }],
    couplingScore: [
      {
        i: 0,
        // tslint:disable-next-line:object-literal-sort-keys
        A_i: 'I',
        j: 1,
        A_j: 'J',
        fn: 1,
        cn: 1,
        segment_i: 'K',
        segment_j: 'L',
        probability: 1,
        dist_intra: 1,
        dist_multimer: 1,
        dist: 10,
        precision: 10,
      },
    ],
    distanceMapMonomer: [{ id: 0, sec_struct_3state: 'A' }],
  };

  test('Should match existing snapshot when given no data.', () => {
    expect(toJson(shallow(<ContactMapComponent />))).toMatchSnapshot();
  });

  test('Should match existing snapshot when given empty data.', () => {
    expect(toJson(shallow(<ContactMapComponent data={emptyData} />))).toMatchSnapshot();
  });

  test('Should match existing snapshot when given sample data.', () => {
    expect(toJson(shallow(<ContactMapComponent data={emptyData} />))).toMatchSnapshot();
  });

  test('Should invoke onClick callback when appropriate.', () => {
    const onClickSpy = jest.fn();
    const wrapper = mount(<ContactMapComponent data={sampleData} onClick={onClickSpy} />);
    wrapper
      .find('.recharts-scatter-symbol')
      .at(0)
      .simulate('click');
  });
});
