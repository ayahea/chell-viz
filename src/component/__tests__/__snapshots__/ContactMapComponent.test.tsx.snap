// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactMapComponent Should match existing snapshot when given empty data. 1`] = `
<div
  style={
    Object {
      "padding": 10,
    }
  }
>
  <ScatterChart
    barCategoryGap="10%"
    barGap={4}
    height={370}
    layout="horizontal"
    margin={
      Object {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    reverseStackOrder={false}
    stackOffset="none"
    width={370}
  >
    <XAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      dataKey="i"
      domain={
        Array [
          1,
          1,
        ]
      }
      height={30}
      hide={false}
      mirror={false}
      orientation="top"
      padding={
        Object {
          "left": 0,
          "right": 0,
        }
      }
      reversed={false}
      scale="auto"
      tickCount={5}
      type="number"
      width={0}
      xAxisId={0}
    />
    <YAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      dataKey="j"
      domain={
        Array [
          1,
          1,
        ]
      }
      height={0}
      hide={false}
      mirror={false}
      orientation="left"
      padding={
        Object {
          "bottom": 0,
          "top": 0,
        }
      }
      reversed={true}
      scale="auto"
      tickCount={5}
      type="number"
      width={60}
      yAxisId={0}
    />
    <Scatter
      animationBegin={0}
      animationDuration={400}
      animationEasing="linear"
      data={Array []}
      fill="#009999"
      hide={false}
      isAnimationActive={true}
      legendType="circle"
      lineJointType="linear"
      lineType="joint"
      name="contacts_monomer"
      onClick={[Function]}
      shape={
        <Dot
          r={4}
        />
      }
      xAxisId={0}
      yAxisId={0}
      zAxisId={0}
    />
    <Scatter
      animationBegin={0}
      animationDuration={400}
      animationEasing="linear"
      data={Array []}
      fill="#000000"
      hide={false}
      isAnimationActive={true}
      legendType="circle"
      lineJointType="linear"
      lineType="joint"
      name="CouplingScoresCompared"
      onClick={[Function]}
      onMouseEnter={[Function]}
      shape={
        <Dot
          r={4}
        />
      }
      xAxisId={0}
      yAxisId={0}
      zAxisId={0}
    />
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={true}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={true}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
  </ScatterChart>
  <ChellSlider
    defaultValue={99}
    label="Probability"
    max={100}
    min={0}
    onChange={[Function]}
  />
  <ChellSlider
    defaultValue={4}
    label="Node Size"
    max={5}
    min={1}
    onChange={[Function]}
  />
</div>
`;

exports[`ContactMapComponent Should match existing snapshot when given no data. 1`] = `
<div
  style={
    Object {
      "padding": 10,
    }
  }
>
  <ScatterChart
    barCategoryGap="10%"
    barGap={4}
    height={370}
    layout="horizontal"
    margin={
      Object {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    reverseStackOrder={false}
    stackOffset="none"
    width={370}
  >
    <XAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      dataKey="i"
      domain={
        Array [
          1,
          1,
        ]
      }
      height={30}
      hide={false}
      mirror={false}
      orientation="top"
      padding={
        Object {
          "left": 0,
          "right": 0,
        }
      }
      reversed={false}
      scale="auto"
      tickCount={5}
      type="number"
      width={0}
      xAxisId={0}
    />
    <YAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      dataKey="j"
      domain={
        Array [
          1,
          1,
        ]
      }
      height={0}
      hide={false}
      mirror={false}
      orientation="left"
      padding={
        Object {
          "bottom": 0,
          "top": 0,
        }
      }
      reversed={true}
      scale="auto"
      tickCount={5}
      type="number"
      width={60}
      yAxisId={0}
    />
    <Scatter
      animationBegin={0}
      animationDuration={400}
      animationEasing="linear"
      data={Array []}
      fill="#009999"
      hide={false}
      isAnimationActive={true}
      legendType="circle"
      lineJointType="linear"
      lineType="joint"
      name="contacts_monomer"
      onClick={[Function]}
      shape={
        <Dot
          r={4}
        />
      }
      xAxisId={0}
      yAxisId={0}
      zAxisId={0}
    />
    <Scatter
      animationBegin={0}
      animationDuration={400}
      animationEasing="linear"
      data={Array []}
      fill="#000000"
      hide={false}
      isAnimationActive={true}
      legendType="circle"
      lineJointType="linear"
      lineType="joint"
      name="CouplingScoresCompared"
      onClick={[Function]}
      onMouseEnter={[Function]}
      shape={
        <Dot
          r={4}
        />
      }
      xAxisId={0}
      yAxisId={0}
      zAxisId={0}
    />
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={true}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={true}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
  </ScatterChart>
  <ChellSlider
    defaultValue={99}
    label="Probability"
    max={100}
    min={0}
    onChange={[Function]}
  />
  <ChellSlider
    defaultValue={4}
    label="Node Size"
    max={5}
    min={1}
    onChange={[Function]}
  />
</div>
`;

exports[`ContactMapComponent Should match existing snapshot when given sample data. 1`] = `
<div
  style={
    Object {
      "padding": 10,
    }
  }
>
  <ScatterChart
    barCategoryGap="10%"
    barGap={4}
    height={370}
    layout="horizontal"
    margin={
      Object {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    reverseStackOrder={false}
    stackOffset="none"
    width={370}
  >
    <XAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      dataKey="i"
      domain={
        Array [
          1,
          1,
        ]
      }
      height={30}
      hide={false}
      mirror={false}
      orientation="top"
      padding={
        Object {
          "left": 0,
          "right": 0,
        }
      }
      reversed={false}
      scale="auto"
      tickCount={5}
      type="number"
      width={0}
      xAxisId={0}
    />
    <YAxis
      allowDataOverflow={false}
      allowDecimals={true}
      allowDuplicatedCategory={true}
      dataKey="j"
      domain={
        Array [
          1,
          1,
        ]
      }
      height={0}
      hide={false}
      mirror={false}
      orientation="left"
      padding={
        Object {
          "bottom": 0,
          "top": 0,
        }
      }
      reversed={true}
      scale="auto"
      tickCount={5}
      type="number"
      width={60}
      yAxisId={0}
    />
    <Scatter
      animationBegin={0}
      animationDuration={400}
      animationEasing="linear"
      data={Array []}
      fill="#009999"
      hide={false}
      isAnimationActive={true}
      legendType="circle"
      lineJointType="linear"
      lineType="joint"
      name="contacts_monomer"
      onClick={[Function]}
      shape={
        <Dot
          r={4}
        />
      }
      xAxisId={0}
      yAxisId={0}
      zAxisId={0}
    />
    <Scatter
      animationBegin={0}
      animationDuration={400}
      animationEasing="linear"
      data={Array []}
      fill="#000000"
      hide={false}
      isAnimationActive={true}
      legendType="circle"
      lineJointType="linear"
      lineType="joint"
      name="CouplingScoresCompared"
      onClick={[Function]}
      onMouseEnter={[Function]}
      shape={
        <Dot
          r={4}
        />
      }
      xAxisId={0}
      yAxisId={0}
      zAxisId={0}
    />
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={true}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={true}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
  </ScatterChart>
  <ChellSlider
    defaultValue={99}
    label="Probability"
    max={100}
    min={0}
    onChange={[Function]}
  />
  <ChellSlider
    defaultValue={4}
    label="Node Size"
    max={5}
    min={1}
    onChange={[Function]}
  />
</div>
`;
